### PR TITLE
INT: Remove dbg!

### PR DESCRIPTION
**Pull request by gfreezy**
Remove `dbg` macro.

### CODE REVIEW

File: src/main/resources/intentionDescriptions/RemoveDbgIntention/description.html [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<html>
+<body>
+This intention removes the dbg! call.
```
Comment (ortem) [inline src/main/resources/intentionDescriptions/RemoveDbgIntention/description.html]:
```suggestion
This intention removes the <code>dbg!</code> call.
```

File: src/main/kotlin/org/rust/ide/intentions/RemoveDbgIntention.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * ```
+ */
+class RemoveDbgIntention : RsElementBaseIntentionAction<RsMacroExpr>() {
+
+    override fun getText() = "Remove dbg!"
+    override fun getFamilyName() = text
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): RsMacroExpr? {
+        val macroExpr = element.ancestorStrict<RsMacroExpr>() ?: return null
+        if (macroExpr.macroCall.path.text != "dbg") {
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/intentions/RemoveDbgIntention.kt]:
```suggestion
        if (macroExpr.macroCall.macroName != "dbg") {
```

File: src/test/kotlin/org/rust/ide/intentions/RemoveDbgIntentionTest.kt [hunk 3]
```diff
@@ -0,0 +1,30 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.intentions
+
+
+class RemoveDbgIntentionTest : RsIntentionTestBase(RemoveDbgIntention()) {
```
Comment (ortem) [inline src/test/kotlin/org/rust/ide/intentions/RemoveDbgIntentionTest.kt]: Please provide more tests. For example, with something like `foo(x, dbg!((5 + 7) * 2), y)`

Comment (gfreezy) [reply src/test/kotlin/org/rust/ide/intentions/RemoveDbgIntentionTest.kt]: 👌

Comment (gfreezy) [reply src/test/kotlin/org/rust/ide/intentions/RemoveDbgIntentionTest.kt]: I have no idea why `caret` will be in this position after remove `dbg!`.

```kotlin
    fun `test remove outer dbg!`() = doAvailableTest("""
        fn test() {
            let a = dbg!(1 +/*caret*/ dbg!(3));
        }
    """, """
        fn test() {
            let a = 1 + dbg!/*caret*/(3);
        }
    """)
```

Comment (ortem) [reply src/test/kotlin/org/rust/ide/intentions/RemoveDbgIntentionTest.kt]: You should move caret after replace, and don't forget to invoke `doPostponedOperationsAndUnblockDocument`:
```Kotlin
override fun invoke(project: Project, editor: Editor, ctx: RsMacroExpr) {
    val expr = ctx.macroCall.exprMacroArgument?.expr ?: return
    val newExpr = ctx.replace(expr)
    PsiDocumentManager.getInstance(project).doPostponedOperationsAndUnblockDocument(editor.document)
    editor.caretModel.moveToOffset(newExpr.textOffset)
}
```

Comment (ortem) [reply src/test/kotlin/org/rust/ide/intentions/RemoveDbgIntentionTest.kt]: Also, please squash your commits into one before merge

File: src/main/kotlin/org/rust/ide/intentions/RemoveDbgIntention.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        val macroExpr = element.ancestorStrict<RsMacroExpr>() ?: return null
+        if (macroExpr.macroCall.macroName != "dbg") {
+            return null
+        }
+        return macroExpr
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: RsMacroExpr) {
+        val expr = ctx.macroCall.exprMacroArgument?.expr ?: return
+        val cursorOffsetToExpr = editor.caretModel.offset - expr.startOffset
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/intentions/RemoveDbgIntention.kt]: Looks like it doesn't work in such cases:
`foobar(d/*caret*/bg!(123));`
`foobar(dbg!(123   <long whitespace>   /*caret*/));`

Try use `kotlin.math.min/max` like this:
```
val cursorOffsetToExpr = max(0, editor.caretModel.offset - expr.startOffset)
val resultOffset = min(newExpr.startOffset + cursorOffsetToExpr, newExpr.endOffset)
```
and add such tests

**Comment by gfreezy** [comment]
@ortem Code updated

**Comment by ortem** [comment]
@gfreezy Thank you!
bors r+

**Comment by Undin** [comment]
bors retry

**Pull request merged**